File: Assets/Bundles/Shared/Resources/TS/Util/Spring.lua
-- Compiled with unity-ts v2.1.0-75
local Spring
do
	Spring = setmetatable({}, {
		__tostring = function()
			return "Spring"
		end,
	})
	Spring.__index = Spring
	function Spring.new(...)
		local self = setmetatable({}, Spring)
		return self:constructor(...) or self
	end
	function Spring:constructor(position, angularFrequency)
		self.position = position
		self.angularFrequency = angularFrequency
		self.goal = position
		self.velocity = Vector3.new(0, 0, 0)
	end
	function Spring:update(deltaTime)
		local angFreq = self.angularFrequency * math.pi * 2
		local goal = self.goal
		local p0 = self.position
		local v0 = self.velocity
		local offset = p0 - goal
		local decay = math.exp(-angFreq * deltaTime)
		local _arg0 = 1 + angFreq * deltaTime
		local _exp = offset * _arg0
		local _deltaTime = deltaTime
		local _arg0_1 = v0 * _deltaTime
		local position = (_exp + _arg0_1) * decay + goal
		local _arg0_2 = 1 - angFreq * deltaTime
		local _exp_1 = v0 * _arg0_2
		local _arg0_3 = angFreq * angFreq * deltaTime
		local _arg0_4 = offset * _arg0_3
		self.velocity = (_exp_1 - _arg0_4) * decay
		self.position = position
		return position
	end
	function Spring:resetTo(position)
		self.position = position
		self.goal = position
		self.velocity = Vector3.new(0, 0, 0)
	end
	function Spring:impulse(impulse)
		local _velocity = self.velocity
		local _impulse = impulse
		self.velocity = _velocity + _impulse
	end
end
return {
	Spring = Spring,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Bundles/@Easy/Core/Core~/src/Shared/Util/Spring.ts
export class Spring {
	/** The goal for the spring. */
	public goal: Vector3;

	/** The current velocity of the spring. */
	public velocity: Vector3;

	constructor(
		/** The position of the spring */
		private position: Vector3,

		/** The angular frequency of the spring. */
		private angularFrequency: number,
	) {
		this.goal = position;
		this.velocity = new Vector3(0, 0, 0);
	}

	/** Update the spring. */
	public Update(deltaTime: number): Vector3 {
		const angFreq = this.angularFrequency * math.pi * 2;
		const goal = this.goal;
		const p0 = this.position;
		const v0 = this.velocity;
		const offset = p0.sub(goal);
		const decay = math.exp(-angFreq * deltaTime);
		const position = offset
			.mul(1 + angFreq * deltaTime)
			.add(v0.mul(deltaTime))
			.mul(decay)
			.add(goal);
		this.velocity = v0
			.mul(1 - angFreq * deltaTime)
			.sub(offset.mul(angFreq * angFreq * deltaTime))
			.mul(decay);
		this.position = position;
		return position;
	}

	/** Reset the spring to a specific position. */
	public ResetTo(position: Vector3) {
		this.position = position;
		this.goal = position;
		this.velocity = new Vector3(0, 0, 0);
	}

	/** Impulse the spring with a given impulse velocity. */
	public Impulse(impulse: Vector3) {
		this.velocity = this.velocity.add(impulse);
	}
}
